docs/_specs/knative-build/knative-build.md

```diff
@@ -0,0 +1,204 @@
+
+## Proposal
+
+Create two build templates: one for images and one for functions. The build template will utilize `kaniko` to build the docker image and push it to the given registry. We can utilize the existing base images we use now without any changes.
+
+## Images
+
+### ***Image Build Template***
+
+Here is a possible image template:
+```yaml
+apiVersion: build.knative.dev/v1alpha1
+kind: BuildTemplate
+metadata:
+  name: image-template
+  namespace: dispatch
+spec:
+  parameters:
+  - name: DESTINATION
+    description: The destination to push the image (image name)
+  - name: BASE_IMAGE
+    description: The base image which this image is built from
+  - name: DOCKERFILE
+    description: Path to the Dockerfile to build
+    default: Dockerfile
+  - name: SYSTEM_PACKAGES_FILE
+    description: Path to file with system dependencies
+    default: system-packages.txt
+  - name: PACKAGES_FILE
+    description: Path to file with runtime dependencies
+    default: packages.txt
+
+  steps:
+  - name: build-and-push
+    image: gcr.io/kaniko-project/executor:v0.3.0
+    args:
+    - --dockerfile=${DOCKERFILE}
+    - --destination=${DESTINATION}
+    #- --insecure-skip-tls-verify (not yet implemented)
+    - --build-arg=BASE_IMAGE=${BASE_IMAGE}
+    - --build-arg=SYSTEM_PACKAGES_FILE=${SYSTEM_PACKAGES_FILE}
+    - --build-arg=PACKAGES_FILE=${PACKAGES_FILE}
+```
+
+This build template accepts multiple arguments that will be used to configure the build. The `DESTINATION` and `BASE_IMAGE` arguments are required. `Kaniko` expects the docker context (`Dockerfile`, `system-packages.txt`, `packages.txt`) to be in the `/workspace` directory, which should be provided by the sources in the knative build.
+
+Insecure registries for kaniko are not supported yet. It is currently [WIP](https://github.com/GoogleContainerTools/kaniko/pull/169).
+
+### ***Image Build Example***
+
+When creating an image, Dispatch will be responsible for using the build template to create a Knative build. Dispatch should use the Knative build api to construct a Knative build object with the build template and send it to kubernetes to start the build.
+
+For example, for a `nodejs` image, it may construct a build object with the given yaml:
+
+```yaml
+apiVersion: build.knative.dev/v1alpha1
+kind: Build
+metadata:
+  name: nodejs
+  namespace: dispatch
+spec:
+  #serviceAccountName: docker-reg-auth
+  source:
+    custom:
+      image: dispatchframework/nodejs-base:0.0.12 # base-image
+      command: ['cp']
+      args: ['${IMAGE_TEMPLATE_DIR}/Dockerfile', '/workspace']
+      env:
+      - name: 'IMAGE_TEMPLATE_DIR'
+        value: /image-template
+    custom:
+      image: echo # see echo image
+      args: ['-n', '${SYSTEM_PACKAGES_CONTENT}', '|',  'base64', '--decode', '>', '/workspace/${SYSTEM_PACKAGES_FILE}']
```

Comment:
Writing System Dependencies & Packages could be a build step in the `buildTemplate` rather than a part of the `Build` source.